### PR TITLE
L0 for app service Deploy task

### PR DESCRIPTION
**Pull request by RoshanKumarMicrosoft**

### CODE REVIEW

**Comment by rokmicrosoft** [comment]
Still adding More L0.

File: Tasks/AzureRmWebAppDeploymentV4/Tests/AzureRmWebAppDeploymentProviderL0Tests.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+import { getMockEndpoint } from '../node_modules/azure-arm-rest/Tests/mock_utils';
+import { mockAzureARMPreDeploymentSteps }  from "./mock_utils";
+
+getMockEndpoint();
+mockAzureARMPreDeploymentSteps();
+
+export class AzureRmWebAppDeploymentProviderL0Tests  {
+
+    public static async startAzureRmWebAppDeploymentProviderL0Tests() {
+        await AzureRmWebAppDeploymentProviderL0Tests.testForPreDeploymentSteps();
```
Comment (SumiranAgg) [inline Tasks/AzureRmWebAppDeploymentV4/Tests/AzureRmWebAppDeploymentProviderL0Tests.ts]: Should we add tests for virtual application and post deployment step

File: Tasks/AzureRmWebAppDeploymentV4/Tests/WindowsWebAppRunFromZipProviderL0Tests.ts [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+getMockEndpoint();
+mockAzureARMPreDeploymentSteps();
+mockRunFromZipSettings();
+
+export class WindowsWebAppRunFromZipProviderL0Tests  {
+
+    public static async startWindowsWebAppRunFromZipProviderL0Tests() {
+        await WindowsWebAppRunFromZipProviderL0Tests.testForPreDeploymentSteps_RunFromZipProvider();
+        await WindowsWebAppRunFromZipProviderL0Tests.testForPreDeploymentStepsWithSlotEnabled_RunFromZipProvider();
```
Comment (SumiranAgg) [inline Tasks/AzureRmWebAppDeploymentV4/Tests/WindowsWebAppRunFromZipProviderL0Tests.ts]: with folder package type

File: Tasks/AzureRmWebAppDeploymentV4/Tests/DeploymentFactoryL0Tests.ts [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+import { IWebAppDeploymentProvider } from '../deploymentProvider/IWebAppDeploymentProvider';
+import { TaskParametersUtility, TaskParameters, DeploymentType } from '../operations/TaskParameters';
+import { stringify } from 'querystring';
+import { PackageType } from 'webdeployment-common/packageUtility';
+
+export class DeploymentFactoryL0Tests  {
+
+    public static async startDeploymentFactoryL0Tests() {
+        await DeploymentFactoryL0Tests.testForLinuxWebDeploymentProvider();
+        await DeploymentFactoryL0Tests.testForWindowsWebAppRunFromZipProvider();
```
Comment (SumiranAgg) [inline Tasks/AzureRmWebAppDeploymentV4/Tests/DeploymentFactoryL0Tests.ts]: publishprofileprovider, container provider
Also can we cover providers for _getUserSelectedDeploymentProviderForWindow and _getWindowsDeploymentProviderForZipAndFolderPackageType
As discussed offline just trying to maintain all if conditions are present even if in future someone refactors.

File: Tasks/AzureRmWebAppDeploymentV4/Tests/WindowsWebAppWebDeployProviderL0Tests.ts [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+getMockEndpoint();
+mockAzureARMPreDeploymentSteps();
+mockZipDeploySettings();
+
+export class WindowsWebAppWebDeployProviderL0Tests  {
+
+    public static async startWindowsWebAppWebDeployProviderL0Tests() {
+        await WindowsWebAppWebDeployProviderL0Tests.testForPreDeploymentSteps_WebDeployProvider();
+        await WindowsWebAppWebDeployProviderL0Tests.testForPreDeploymentStepsWithSlotEnabled_WebDeployProvider();
+        await WindowsWebAppWebDeployProviderL0Tests.testForUpdateDeploymentStatus_WebDeployProvider();
```
Comment (SumiranAgg) [inline Tasks/AzureRmWebAppDeploymentV4/Tests/WindowsWebAppWebDeployProviderL0Tests.ts]: Please add for virtual application and when MSDeploy package type

File: Tasks/AzureRmWebAppDeploymentV4/Tests/WindowsWebAppZipDeployProviderL0Tests.ts [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+getMockEndpoint();
+mockAzureARMPreDeploymentSteps();
+mockZipDeploySettings();
+
+export class WindowsWebAppZipDeployProviderL0Tests  {
+
+    public static async startWindowsWebAppZipDeployProviderL0Tests() {
+        await WindowsWebAppZipDeployProviderL0Tests.testForPreDeploymentSteps_ZipDeployProvider();
+        await WindowsWebAppZipDeployProviderL0Tests.testForPreDeploymentStepsWithSlotEnabled_ZipDeployProvider();
```
Comment (SumiranAgg) [inline Tasks/AzureRmWebAppDeploymentV4/Tests/WindowsWebAppZipDeployProviderL0Tests.ts]: add for folder package type

File: Tasks/AzureRmWebAppDeploymentV4/Tests/AzureRmWebAppDeploymentProviderTests.ts [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+import { AzureEndpoint } from 'azure-arm-rest/azureModels';
+import { ApplicationTokenCredentials } from 'azure-arm-rest/azure-arm-common';
+import { AzureRMEndpoint } from 'azure-arm-rest/azure-arm-endpoint'; 
+import { setEndpointData, setAgentsData, mockTaskArgument } from './utils';
+
+export class AzureRmWebAppDeploymentProviderTests {
+
+    public static startAzureRmWebAppDeploymentProviderTests(){
+        let tp = path.join(__dirname, 'AzureRmWebAppDeploymentProviderL0Tests.js');
+        let tr : tmrm.TaskMockRunner = new tmrm.TaskMockRunner(tp);
```
Comment (SumiranAgg) [inline Tasks/AzureRmWebAppDeploymentV4/Tests/AzureRmWebAppDeploymentProviderTests.ts]: Can we have a common tr object with default values and override wherever necessary. I see similar object in all files, we can remove some redundancy.

File: Tasks/AzureRmWebAppDeploymentV4/Tests/AzureRmWebAppDeploymentProviderL0Tests.ts (hunk 1 above)
Comment (RoshanKumarMicrosoft) [reply Tasks/AzureRmWebAppDeploymentV4/Tests/AzureRmWebAppDeploymentProviderL0Tests.ts]: This is base class and for each subsequent extended class post deployment step is a part of DeployWebAppStep. Already added corresponding L0 for each subsequent extended class.  For virtual application I am adding L0.

File: Tasks/AzureRmWebAppDeploymentV4/Tests/AzureRmWebAppDeploymentProviderTests.ts (hunk 6 above)
Comment (RoshanKumarMicrosoft) [reply Tasks/AzureRmWebAppDeploymentV4/Tests/AzureRmWebAppDeploymentProviderTests.ts]: Moved into common utils through this commit https://github.com/Microsoft/vsts-tasks/pull/8027/commits/bb951457d8f8ae91b3b62faec2bde2cdd4f9787f

File: Tasks/AzureRmWebAppDeploymentV4/Tests/ContainerWebAppDeploymentProviderL0Tests.ts [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+getMockEndpoint();
+mockAzureARMPreDeploymentSteps();
+
+export class ContainerWebAppDeploymentProviderL0Tests  {
+
+    public static async startContainerWebAppDeploymentProviderL0Tests() {
+        await ContainerWebAppDeploymentProviderL0Tests.testForPreDeploymentSteps_ContainerWebApp();
+        await ContainerWebAppDeploymentProviderL0Tests.testForPreDeploymentStepsWithSlotEnabled_ContainerWebApp();
+        await ContainerWebAppDeploymentProviderL0Tests.testForUpdateDeploymentStatus_ContainerWebApp();
```
[the remark this answers is not in the document]
Comment (RoshanKumarMicrosoft) [reply Tasks/AzureRmWebAppDeploymentV4/Tests/ContainerWebAppDeploymentProviderL0Tests.ts]: Done . Added  on this commit 6d133d3f61515e5410119960b98df7f778e654b7

File: Tasks/AzureRmWebAppDeploymentV4/Tests/ContainerWebAppDeploymentProviderTests.ts [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    public static startContainerWebAppDeploymentProviderTests(){
+        let tp = path.join(__dirname, 'ContainerWebAppDeploymentProviderL0Tests.js');
+        let tr : tmrm.TaskMockRunner = new tmrm.TaskMockRunner(tp);
+        tr.setInput("ConnectionType", "AzureRM");
+        tr.setInput('ConnectedServiceName', 'AzureRMSpn');
+        tr.setInput('WebAppName', 'mytestapp');
+        tr.setInput('Package', 'webAppPkg.zip');
+        tr.setInput('UseWebDeploy', 'false');
+        tr.setInput('ImageSource', "Builtin Image");
+        tr.setInput('WebAppKind', "webAppLinux");
```
[the remark this answers is not in the document]
Comment (RoshanKumarMicrosoft) [reply Tasks/AzureRmWebAppDeploymentV4/Tests/ContainerWebAppDeploymentProviderTests.ts]: Now it is in common and it has collection input of all dependent class.

File: Tasks/AzureRmWebAppDeploymentV4/Tests/DeploymentFactoryL0Tests.ts (hunk 3 above)
Comment (RoshanKumarMicrosoft) [reply Tasks/AzureRmWebAppDeploymentV4/Tests/DeploymentFactoryL0Tests.ts]: Done. Added in this commit 55eb84f00332369924f7fbb5891bbc0380ea5d22

File: Tasks/AzureRmWebAppDeploymentV4/Tests/WindowsWebAppZipDeployProviderL0Tests.ts (hunk 5 above)
Comment (RoshanKumarMicrosoft) [reply Tasks/AzureRmWebAppDeploymentV4/Tests/WindowsWebAppZipDeployProviderL0Tests.ts]: Done. Added in this commit fdd57b62de16f1d6f2671cd79cf1988304381151

File: Tasks/AzureRmWebAppDeploymentV4/Tests/WindowsWebAppRunFromZipProviderL0Tests.ts (hunk 2 above)
Comment (RoshanKumarMicrosoft) [reply Tasks/AzureRmWebAppDeploymentV4/Tests/WindowsWebAppRunFromZipProviderL0Tests.ts]: Done. Added in this commit fdd57b62de16f1d6f2671cd79cf1988304381151

File: Tasks/AzureRmWebAppDeploymentV4/Tests/WindowsWebAppWebDeployProviderL0Tests.ts (hunk 4 above)
Comment (RoshanKumarMicrosoft) [reply Tasks/AzureRmWebAppDeploymentV4/Tests/WindowsWebAppWebDeployProviderL0Tests.ts]: Done. In this commit 6799bec05dd4d567b32c2661ea25a6282f03adf3

**Pull request merged**